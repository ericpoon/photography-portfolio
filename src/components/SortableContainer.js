import React from 'react';
import { DragSource, DropTarget } from 'react-dnd';
import flow from 'lodash/flow';

const itemType = 'SORTABLE';

class SortableContainer extends React.Component {
  render() {
    const { connectDragSource, connectDropTarget, children } = this.props;
    return connectDragSource(connectDropTarget(
      <div
        ref={(node) => {
          this.node = node;
        }}
      >
        {children}
      </div>
    ));
  }
}

const cardSource = {
  beginDrag(props) {
    let { index } = props;
    return {
      get index() {
        return index;
      },
      setIndex(newIndex) {
        index = newIndex;
      },
    };
  },
};

const cardTarget = {
  hover(props, monitor, component) {
    const { decoratedComponentInstance } = component;
    const dragIndex = monitor.getItem().index;
    const hoverIndex = props.index;

    if (dragIndex === hoverIndex) return;

    // Determine rectangle on screen, avoid using findDOMNode
    const hoverBoundingRect = decoratedComponentInstance.node.getBoundingClientRect();

    // Determine mouse position
    const clientOffset = monitor.getClientOffset();

    if (clientOffset.x > hoverBoundingRect.left
      && clientOffset.x < hoverBoundingRect.right
      && clientOffset.y > hoverBoundingRect.top
      && clientOffset.y < hoverBoundingRect.bottom) {
      props.trigger(dragIndex, hoverIndex); // actually perform the action
      monitor.getItem().setIndex(hoverIndex); // update the item being dragged
    }
  },
};

const sourceCollect = (connect, monitor) => ({
  connectDragSource: connect.dragSource(),
  isDragging: monitor.isDragging(),
});
const targetCollect = (connect, monitor) => ({
  connectDropTarget: connect.dropTarget(),
});

export default flow(
  DragSource(itemType, cardSource, sourceCollect),
  DropTarget(itemType, cardTarget, targetCollect)
)(SortableContainer);
