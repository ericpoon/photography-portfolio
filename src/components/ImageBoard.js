import React from 'react';
import { DragDropContext } from 'react-dnd';
import HTML5Backend from 'react-dnd-html5-backend';
import SortableContainer from './SortableContainer';

const ROW_NUM = 3;
const COL_NUM = 3;

// todo: Item states are not persisted,
// use redux store to keep track of those images being edited (not all)

class ImageBoard extends React.Component {
  constructor(props) {
    super();
    this.state = { images: props.children };
  }

  componentWillReceiveProps(nextProps, nextContext) {
    this.setState({ images: nextProps.children });
  }

  moveCard = (dragIndex, hoverIndex) => {
    const { images } = this.state;
    const draggedCard = images[dragIndex];
    images.splice(dragIndex, 1, images[hoverIndex]);
    images.splice(hoverIndex, 1, draggedCard);
    this.setState({ images });
  };

  render() {
    const { images } = this.state;
    const { sortable } = this.props;
    const grid = [];

    images.forEach((current, index) => {
      const row = Math.floor(index / ROW_NUM);
      const col = index % COL_NUM;
      if (!grid[row]) grid[row] = [];
      grid[row][col] = sortable ? (
        <SortableContainer
          moveCard={this.moveCard}
          index={index}
          key={current.key}
        >
          {current}
        </SortableContainer>
      ) : current;
    });

    return (
      <div className={'image-card-collection'}>
        {grid.map(row => (
          <div className={'image-card-collection__row'} key={row.map(i => i.key).toString()}>
            {row.map(individual => individual)}
          </div>
        ))}
      </div>
    );
  }
}

export default DragDropContext(HTML5Backend)(ImageBoard);
