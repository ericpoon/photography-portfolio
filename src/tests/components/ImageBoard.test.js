import React from 'react';
import { mount } from 'enzyme';
import TestBackend from 'react-dnd-test-backend';
import { DragDropContext } from 'react-dnd';
import SortableContainer from '../../components/SortableContainer';
import ImageBoard, { ImageBoard as UnconnectedImageBoard } from '../../components/ImageBoard';

describe('Testing the original component in isolation', () => {
  let wrapper;
  beforeEach(() => {
    const MockImageCard = jest.fn(() => (<div />));
    wrapper = mount(
      <ImageBoard sortable>
        <MockImageCard key={0} />
        <MockImageCard key={1} />
        <MockImageCard key={2} />
      </ImageBoard>);
  });

  it('should render ImageBoard', () => {
    wrapper.setProps({ sortable: false });
    expect(wrapper).toMatchSnapshot();
  });

  it('should render sortable ImageBoard', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should wrap children within SortableContainer', () => {
    expect(wrapper.find(SortableContainer)).toHaveLength(3);
  });
});

describe('Testing the drag and drop interaction', () => {
  /* eslint-disable react/prefer-stateless-function */
  const wrapInTestContext = (DecoratedComponent) => {
    return DragDropContext(TestBackend)(
      // have to make this a class component in order to access ref
      class TestContextContainer extends React.Component {
        render() {
          return <DecoratedComponent {...this.props} />;
        }
      }
    );
  };
  /* eslint-enable */

  let wrapper, manager, backend, monitor, sourceId, targetId, targetNode;
  beforeEach(() => {
    const Wrapped = wrapInTestContext(UnconnectedImageBoard); // must be unconnected

    wrapper = mount(
      <Wrapped sortable>
        <div key={1} />
        <div key={2} />
      </Wrapped>
    );

    manager = wrapper.instance().getManager();
    backend = manager.getBackend();
    monitor = manager.getMonitor();

    const dragSource = wrapper.find(SortableContainer).at(0).find('DragSource(SortableContainer)').instance();
    const dropTarget = wrapper.find(SortableContainer).at(1).instance();

    sourceId = dragSource.getHandlerId();
    targetId = dropTarget.getHandlerId();
    targetNode = dropTarget
      .decoratedComponentInstance
      .decoratedComponentInstance
      .node;
  });

  it('should not trigger reorder', () => {
    monitor.getClientOffset = jest.fn(() => ({
      x: 1000,
      y: 1000,
    }));

    targetNode.getBoundingClientRect = jest.fn(() => ({
      left: 100,
      right: 250,
      top: 0,
      bottom: 100,
      width: 150,
      height: 100,
    }));

    backend.simulateBeginDrag([sourceId]);
    backend.simulateHover([targetId]);
  });

  it('should trigger reorder', () => {
    monitor.getClientOffset = jest.fn(() => ({
      x: 150,
      y: 50,
    }));

    targetNode.getBoundingClientRect = jest.fn(() => ({
      left: 100,
      right: 250,
      top: 0,
      bottom: 100,
      width: 150,
      height: 100,
    }));

    backend.simulateBeginDrag([sourceId]);
    // backend.simulateHover([targetId]);
    // expect(mockTrigger).toHaveBeenCalledTimes(1);
    /* FIXME
    * should somehow mock trigger function in props
    * and expect it to be called once
    * probably tell backend.simulateHover to mock props.trigger
    * */
  });
});
