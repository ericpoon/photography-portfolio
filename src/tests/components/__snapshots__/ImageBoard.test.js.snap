// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the original component in isolation should render ImageBoard 1`] = `
<DragDropContext(ImageBoard)
  sortable={false}
>
  <ImageBoard
    sortable={false}
  >
    <div
      className="image-card-collection"
    >
      <div
        className="image-card-collection__row"
        key="0,1,2"
      >
        <mockConstructor
          key="0"
        >
          <div />
        </mockConstructor>
        <mockConstructor
          key="1"
        >
          <div />
        </mockConstructor>
        <mockConstructor
          key="2"
        >
          <div />
        </mockConstructor>
      </div>
    </div>
  </ImageBoard>
</DragDropContext(ImageBoard)>
`;

exports[`Testing the original component in isolation should render sortable ImageBoard 1`] = `
<DragDropContext(ImageBoard)
  sortable={true}
>
  <ImageBoard
    sortable={true}
  >
    <div
      className="image-card-collection"
    >
      <div
        className="image-card-collection__row"
        key="0,1,2"
      >
        <DropTarget(DragSource(SortableContainer))
          index={0}
          key="0"
          trigger={[Function]}
        >
          <DragSource(SortableContainer)
            connectDropTarget={[Function]}
            index={0}
            trigger={[Function]}
          >
            <SortableContainer
              connectDragSource={[Function]}
              connectDropTarget={[Function]}
              index={0}
              isDragging={false}
              trigger={[Function]}
            >
              <div>
                <mockConstructor
                  key="0"
                >
                  <div />
                </mockConstructor>
              </div>
            </SortableContainer>
          </DragSource(SortableContainer)>
        </DropTarget(DragSource(SortableContainer))>
        <DropTarget(DragSource(SortableContainer))
          index={1}
          key="1"
          trigger={[Function]}
        >
          <DragSource(SortableContainer)
            connectDropTarget={[Function]}
            index={1}
            trigger={[Function]}
          >
            <SortableContainer
              connectDragSource={[Function]}
              connectDropTarget={[Function]}
              index={1}
              isDragging={false}
              trigger={[Function]}
            >
              <div>
                <mockConstructor
                  key="1"
                >
                  <div />
                </mockConstructor>
              </div>
            </SortableContainer>
          </DragSource(SortableContainer)>
        </DropTarget(DragSource(SortableContainer))>
        <DropTarget(DragSource(SortableContainer))
          index={2}
          key="2"
          trigger={[Function]}
        >
          <DragSource(SortableContainer)
            connectDropTarget={[Function]}
            index={2}
            trigger={[Function]}
          >
            <SortableContainer
              connectDragSource={[Function]}
              connectDropTarget={[Function]}
              index={2}
              isDragging={false}
              trigger={[Function]}
            >
              <div>
                <mockConstructor
                  key="2"
                >
                  <div />
                </mockConstructor>
              </div>
            </SortableContainer>
          </DragSource(SortableContainer)>
        </DropTarget(DragSource(SortableContainer))>
      </div>
    </div>
  </ImageBoard>
</DragDropContext(ImageBoard)>
`;
